include/dqlite/client.h

```diff
@@ -0,0 +1,361 @@
+ * declared address is the address of this node; #DQLITE_CLUSTER_BOOTSTRAP,
+ * indicating that the node with the declared address will be the first
+ * to run when the cluster starts up for the first time. Exactly one address
+ * should be declared with #DQLITE_CLUSTER_BOOTSTRAP, and the same boostrap
+ * address should be declared on all nodes.
+ *
+ * No reference to the string pointed to by @addr is kept after this function
+ * returns.
+ */
+int dqlite_cluster_declare_node(dqlite_cluster *cluster,
+				const char *addr,
+				int flags);
+
+/**
+ * Set the address on which this instance's server will listen for requests
+ * from client and messages from other servers.
+ *
+ * The format of the @addr argument must be one of
+ *
+ * 1. "<HOST>"
+ * 2. "<HOST>:<PORT>"
+ * 3. "@<PATH>"
+ *
+ * Where <HOST> is a numeric IPv4/IPv6 address, <PORT> is a port number, and
+ * <PATH> is an abstract Unix socket path. The port number defaults to 8080 if
+ * not specified. In the second form, if <HOST> is an IPv6 address, it must be
+ * enclosed in square brackets "[]". In the third form, if <PATH> is empty, the
+ * implementation will automatically select an available abstract Unix socket
+ * path, which can then be retrieved with dqlite_node_get_bind_address().
+ *
+ * If an abstract Unix socket is used the dqlite node will accept only
+ * connections originating from the same process.
+ *
+ * No reference to the memory pointed to by @address is kept, so any memory
+ * associated with them can be released after the function returns.
+ */
+int dqlite_cluster_set_server_bind_address(dqlite_cluster *cluster,
+					   const char *addr);
+
+/**
+ * Customize the function that this instance will use to connect to servers.
+ *
+ * The same callback will be used by this instance's client to contact servers,
+ * and by the instance's server to contact other servers. The callback may be
+ * invoked concurrently on multiple threads, and is responsible for handling
+ * any required synchronization internally.
+ */
+int dqlite_cluster_set_connect_func(dqlite_cluster *cluster,
+				    dqlite_connect_func connect,
+				    void *connect_arg);
+
+/**
+ * Configure the target number of voting nodes for this cluster.
+ *
+ * This must be an odd number.
+ */
+int dqlite_cluster_set_voters(dqlite_cluster *cluster, unsigned n);
+
+/**
+ * Configure the target number of standby nodes for this cluster.
+ */
+int dqlite_cluster_set_standbys(dqlite_cluster *cluster, unsigned n);
+
+/**
+ * Launch the client and server for this instance, attempting to join the
+ * configured cluster.
+ */
+int dqlite_cluster_start(dqlite_cluster *cluster);
+
+/**
+ * Open a database connection on the given cluster.
+ *
+ * This is the analogue of sqlite3_open. @name is the name of the database
+ * to open, which will be created if it does not exist. All nodes in the cluster
+ * share a "namespace" for databases.
+ */
+int dqlite_open(dqlite_cluster *cluster, const char *name, dqlite **db);
```

Comment:
Maybe we could rename this one to `dqlite_cluster_client()`?

The reason is that we might want to leave the `dqlite_open` name for a future lower-level API, an API that does not need a `dqlite_cluster` (and `dqlite_cluster_client()` would be basically a helper around `dqlite_open()`, which also knows about current cluster addresses).

Users should be able to use `dqlite_open()` to create client a "pure" dqlite client, without any underlying `dqlite_node` (think for example implementing a dqlite shell in C). The parameters you would pass to `dqlite_open()` could be something like:

```C
dqlite_open(int node_store_type, void* node_store, const char *name, dqlite **db);
```

Where `node_store_type`/`node_store` would be some generic way to pass cluster addresses. Or any other equivalent naming/API.

`dqlite_cluster_client()` could then internally be implemented using `dqlite_open()`.

Reply:
> It's called `dqlite_open` by analogy with `sqlite3_open`, since it was designed to be the direct analogue of that function. But I don't want to spend time debating the name, so I'll change it to dqlite_cluster_cluster if you feel strongly.

I'm saying that we'll need a way to create a client **without** also starting a node, for example if you want to implement a dqlite shell (similar to the current `cmd/dqlite/dqlite.go` program in Go dqlite).

Usually you don't need to run a server in order to create a client, the client can be created independently, and a common naming for creating clients is `open`. We definitely would have `dqlite_open`, analogous to `sqlite3_open`, that's not under debate, but it would be purely a client, without coupling it with the server. I'm feeling strongly about anything, just trying to give my contribution.

Reply:
It's not much effort. You just use a different name and leave the door open for the future. It might come in 3 years, but if the project grows as it should it will, so planning is always good, especially for public APIs.

To be clear, my ones are observation and hints. You're of course the one in charge and you can take them into account or not.